### ISSUE TITLE
[OTBN] Evaluate multiply instructions against different cryptographic use cases

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#2470]
The OTBN v0.1 specification defines a BN.MULQACC instruction, which combines a multiply with an accumulate and shift operation. 

We need to try out using this instruction in the implementation of crypto algorithms to see how it fits in there, especially when used in loops (how much of the loop needs to be unrolled?)

Goal: Figure out if BN.MULQACC works as intended. If changes are needed beyond that, a new issue is needed.

@GregAC has been doing initial investigations.

> estimate 8
> remaining 2020-07-13 2

**Comment by cdgori** [comment]
While not exactly what this issue is covering, probably related because you'll have to try to profile the execution of the crypto algorithms: an actual cycle count for say, a RSA 2048 verify, or a ECC p256 verify would be a really, really helpful thing to be able to measure (either from an instruction set simulator or the current RTL design, once it reaches the point of maturity where it can do such things).

It will help us to understand some things about estimated boot times (@moidx and I were talking about this earlier today), but also to know how OTBN is "doing" in terms of real world performance as we tweak things in the design and evaluate various optimizations.

**Comment by felixmiller** [comment]
You can assign this to me. Will be looking at this when converting the existing algorithms for modular exponentiation and ECDSA to the new ISA

**Comment by GregAC** [comment]
I wrote up a document explaining how the current instruction can be used exploring ways we could extend it if required: https://docs.google.com/document/d/1ZNF-GAfiYqIUiBveCHNw1VW-6im0JFZTPjoW-d3jECA/edit?usp=sharing (I think this should be accessible via any opentitan.org account, if you want access and it's not working let me know and I'll add you manually)

The current issue is the most efficient way to do a multi-limb multiply is a fully unrolled loop. This starts to have a code-size implications as you get bigger multiplies. We could consider a very specialised hardware loop that's entirely for performing schoolbook multiplication using BN.MULQACC to allow larger multiplies that remain efficient without using large amounts of instruction memory.

Another possibility is software will use a series of BN.MULQACC to implement something like 256x256 or 512x512 as efficiently as possible which are then composed to perform larger multiplies. Using the Karatsuba method to do this composition could work well. Alternatively the efficient kernels could be composed using the standard schoolbook algorithm. It wouldn't be efficient as a fully unrolled version but may be good enough.

**Comment by imphil** [comment]
Current status: BN.MULQACC is sufficient for RSA. @felixmiller is looking at ECC next.

**Comment by cdgori** [comment]
Could you comment a little bit more on what "sufficient" means in this context? Contains necessary arithmetic functionality? Meets performance goals?  Something else?  Thanks.

**Comment by felixmiller** [comment]
@cdgori At the moment we're only looking at functional completeness. Constant time is a given though, since we're replacing each MULH instruction with a fixed sequence of multiple instructions.
Main aspect of this issue is to verify that there are no surprises when using such blocks within HW assisted loops. The MULQ instructions do not support indirect addressing of regs, but neither do the MUL128 (dcrypto) or MULH instructions. So, our expectation here is that this goes smooth and it did with the RSA sequence. We still need to verify with the ECC/P256 lib though, to rule out any issues there.

Of course this has other side effects:
- Performance: We need to measure if we still meet our crypto wallclock requirements. @imphil did a lot of research there based on the assumptions how changes in the dcrypto instruction set will translate into OTBN. We need to verify at the end though, with the actual code. This is supposed to be cover in #2635
- Code size: Footprint of the otbn binary is larger with the 64x64 multiplications obviously

**Comment by felixmiller** [comment]
I tested the `MULQACC` instruction now with the ECC lib. [Here](https://github.com/felixmiller/ot-dsim/blob/new-isa/asm/otbn_p256_mul64.asm) is a PoC. Multiplication sequence is [here](https://github.com/felixmiller/ot-dsim/blob/new-isa/asm/otbn_p256_mul64.asm#L65). (This example still contains some dcrypto instructions which is irrelevant at this point, all mult instructions have been replaced)

In principle this works as expected, we have one issue though:

With the 128x128 multiplications (dcrypto or OTBN `MULH`) a typical sequence for a full 256x256 mult looks like this:
```
BN.MULH w3, w1.L, w2.L
BN.MULH w4, w1.U, w2.U
BN.MULH w5, w1.U, w2.L
BN.ADD w3, w3, w5 << 128
BN.ADDC w4, w4, w5 >> 128
BN.MULH w5, w1.L, w2.U
BN.ADD w3, w3, w5 << 128
BN.ADDC w4, w4, w5 >> 128
```

(this gives [w4, w3] = w1*w2)

The last instructions in such a sequence is allways an `ADDC` with the consequence that the flags are set based on the result.
Often the MSB needs to be tested afterwards (e.g. in a subsequent Select instruction).

Workaround in the PoC is to add a [dummy](https://github.com/felixmiller/ot-dsim/blob/new-isa/asm/otbn_p256_mul64.asm#L82) `BN.ADD` at the end of each multiplication and add 0 to the result. This adds quite some overhead though and we should probably set the flags in `BN.MULQACC.SO` and `BN.MULQACC.WO` well.

We need to make sure though that this does not brake things elsewhere (a previously set flag supposed to survive a multiplication instruction)
